### PR TITLE
Add OTOC documentation

### PR DESCRIPTION
**Pull request by rmlarose**
Adds a tutorial based on [this example](https://github.com/quantumlib/ReCirq/blob/144323f212757775a8541dbe2f938b64534b25ec/recirq/otoc/otoc_example.py) which shows how to recreate Fig 1 from the [publication](https://arxiv.org/abs/2101.08870).

### CODE REVIEW

File: docs/otoc/otoc_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "D-6CzQFH84Tr"
+   },
+   "source": [
+    "Quantum scrambling describes how interaction in a quantum system disperses local information into its many degrees of freedom. Analogous to classical chaos, scrambling manifests itself as a [butterfly effect](https://en.wikipedia.org/wiki/Butterfly_effect) wherein a local perturbation is rapidly amplified over time. The perturbation is modeled by a local operator $\\hat{O}$ (the *butterfly operator*) acting on a qubit (the *butterfly qubit*). Under a dynamical process $\\hat{U} = \\hat{U}(t)$, the butterfly operator evolves in the Heisenberg picture as\n",
+    "\n",
+    "$$\n",
+    "\\hat{O} (t) = \\hat{U}^\\dagger \\hat{O} \\hat{U}\n",
```
Comment (mpharrigan) [inline docs/otoc/otoc_example.ipynb:94]: kindof strange that $t$ doesn't appear on the rhs

File: docs/otoc/otoc_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+    "\n",
+    "over a collection of quantum circuits with microscopic differences (e.g., in the phases of gates). Operator scrambling is then reflected in the average OTOC value $\\bar{C}(t)$ over different circuits:\n",
+    "\n",
+    "- $\\bar{C}(t) = 1$ when $\\hat{O}(t)$ and $\\hat{M}$ do not overlap.\n",
+    "- $\\bar{C}(t) < 1$ when $\\hat{O}(t)$ and $\\hat{M}$ overlap.\n",
+    "- $\\bar{C}(t) \\rightarrow 0$ in the fully scrambled limit where the commutation between $\\hat{O}(t)$ and $\\hat{M}$ is completely randomized.\n",
+    "\n",
+    "Note that if operator entanglement is also present ($n_p \\gg 1$), $\\bar{C}(t)$ approaches $0$ for all circuits and their fluctuation vanishes as well.\n",
+    "\n",
+    "In what follows, we show how to build OTOC circuits, run them & compute $\\bar{C}$, then visualize results and detect operator spreading."
```
Comment (mpharrigan) [inline docs/otoc/otoc_example.ipynb:122]: nit why do you use "&" once but "and" in other places

File: docs/otoc/otoc_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,761 @@
...
+    "            print(\"Notebook is not executed with Colab, assuming Application Default Credentials are setup.\") \n",
+    "\n",
+    "    authenticate_user()\n",
+    "    print(\"Successful authentication to Google Cloud.\")\n",
+    "\n",
+    "processor_id = \"\" #@param {type:\"string\"}\n",
+    "if use_noisy_simulator or processor_id == \"\":\n",
+    "    print(\"Using a noisy simulator.\")\n",
+    "    sampler = cg.PhasedFSimEngineSimulator.create_with_random_gaussian_sqrt_iswap(\n",
+    "        mean=cg.SQRT_ISWAP_PARAMETERS,\n",
```
Comment (mpharrigan) [inline docs/otoc/otoc_example.ipynb:262]: no such thing SQRT_ISWAP_PARAMETERS. Do you mean SQRT_ISWAP_INV_PARAMETERS

(non-optional)

File: docs/otoc/otoc_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {
+    "id": "GZIsJoNX1omO"
+   },
+   "outputs": [],
+   "source": [
+    "if not use_noisy_simulator:\n",
```
Comment (mpharrigan) [inline docs/otoc/otoc_example.ipynb:308]: `or processor_id == ""`

[or fix it so `use_noisy_simulator` is True if processor_id == ""]. (non optional)

File: docs/otoc/otoc_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,761 @@
...
+      "(7, 5): ───Y^0.5────────────PhX(-0.25)^0.5───@───PhX(-0.25)^0.5───────────────PhX(0.75)^0.5────@───PhX(0.75)^0.5──────────────────────────────\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(\"Ancilla qubit:\", ancilla)\n",
+    "print(\"Measurement qubit:\", qubits[0])\n",
+    "print(\"Butterfly qubit:\", q_b)\n",
+    "\n",
+    "print(\"\\nExample OTOC circuit:\\n\", otoc_circuits.butterfly_Y[0], sep=\"\\n\")"
```
Comment (mpharrigan) [inline docs/otoc/otoc_example.ipynb:496]: on my jupyter, whenever one does `print(circuit)` it wraps horribly and you can't tell what's going on but if you just have the last line of the cell be `circuit` it uses a horizontal scrollbar and I can actually read it

File: docs/otoc/otoc_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,761 @@
...
+   "metadata": {
+    "id": "zRsJMaG3EkgD"
+   },
+   "outputs": [],
+   "source": [
+    "# Compile to sqrt_iswap.\n",
+    "for i, circuits_trial in enumerate(circuit_list):\n",
+    "    for j, circuits_cycle in enumerate(circuits_trial):\n",
+    "        for k, circuit_otoc in enumerate(circuits_cycle):\n",
+    "            circuit_list[i][j][k] = cg.optimized_for_sycamore(circuit_otoc, optimizer_type=\"sqrt_iswap\")"
```
Comment (mpharrigan) [inline docs/otoc/otoc_example.ipynb:600]: this takes longer than expected on my machine. maybe `print('.', end='')` just so there's some indication of progress

Comment (mpharrigan) [reply docs/otoc/otoc_example.ipynb:600]: this actually takes quite a long time. maybe we can throw multiprocessing at it or have a real progress bar?

**Review by mpharrigan — CHANGES_REQUESTED** [review]
I've noted the "non optional" comments, other comments should be taken under advisement

This is pretty slow, but I guess it comes with the territory. It's also rather unfortunate that we use "meaningful indexing" where you have to sortof note what the various indices mean instead of nice dataclasses, but I understand that's how the code is written and it's outside the scope of this example to change it.

**Comment by mpharrigan** [comment]
should we remove the example script in favor of this notebook?

File: docs/otoc/otoc_example.ipynb (hunk 1 above)
Comment (rmlarose) [reply docs/otoc/otoc_example.ipynb:94]: No change to match with arxiv paper

File: docs/otoc/otoc_example.ipynb (hunk 2 above)
Comment (rmlarose) [reply docs/otoc/otoc_example.ipynb]: Changed & -> and

File: docs/otoc/otoc_example.ipynb (hunk 3 above)
Comment (rmlarose) [reply docs/otoc/otoc_example.ipynb]: Fixed. (1st PR was before Cirq 0.11 update)

File: docs/otoc/otoc_example.ipynb (hunk 4 above)
Comment (rmlarose) [reply docs/otoc/otoc_example.ipynb:311]: Fixed so `use_noisy_simulator` is True if `processor_id == ""`

File: docs/otoc/otoc_example.ipynb (hunk 5 above)
Comment (rmlarose) [reply docs/otoc/otoc_example.ipynb]: Changed to just `circuit` instead of `print(circuit)`

File: docs/otoc/otoc_example.ipynb (hunk 6 above)
Comment (rmlarose) [reply docs/otoc/otoc_example.ipynb]: Added a progress bar. Compiling is kind of a hack around #189 - ideally I think the desired behavior is that each circuit already has the right gates and no compiling is necessary.

**Comment by rmlarose** [comment]
Thanks @mpharrigan! Ready for round 2.

> should we remove the example script in favor of this notebook?

Probably? There's nothing in the example script that isn't in the notebook.

File: docs/otoc/otoc_example.ipynb (hunk 6 above)
Comment (mpharrigan) [reply docs/otoc/otoc_example.ipynb]: ok thanks for the additional information

**Review by mpharrigan — APPROVED** [review]

**Comment by mpharrigan** [comment]
Thanks @rmlarose 

If you want to remove the example script, we can do that in a follow-up PR!

**Pull request merged**